pkg/lib/oauth/handler/handler_token.go

```diff
@@ -207,8 +215,28 @@ func (h *TokenHandler) handleAuthorizationCode(
 		return nil, protocol.NewError("invalid_request", "invalid redirect URI")
 	}
 
-	if codeGrant.PKCEChallenge != "" && !verifyPKCE(codeGrant.PKCEChallenge, r.CodeVerifier()) {
-		return nil, errInvalidAuthzCode
+	// if code verifier is provided, must verify PKCE
+	if r.CodeVerifier() != "" {
+		if codeGrant.PKCEChallenge == "" || !verifyPKCE(codeGrant.PKCEChallenge, r.CodeVerifier()) {
+			return nil, errInvalidAuthzCode
+		}
+	}
+
+	// verify client secret
+	if r.ClientSecret() != "" {
+		credentialsItem, ok := h.OAuthClientCredentials.Lookup(client.ClientID)
+		if !ok {
+			return nil, protocol.NewError("invalid_request", "client secret is not supported for the client")
+		}
+		key, _ := credentialsItem.Set.Get(0)
```

Comment:
To support secret rotation, we should accept any client secret, not just the first one.